// jest-watch-typeahead/build/shared/pattern_mode_helpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatTypeaheadSelection = exports.printTypeaheadItem = exports.printMore = exports.printStartTyping = exports.printRestoredPatternCaret = exports.printPatternCaret = exports.printPatternMatches = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _ansiEscapes = require('ansi-escapes');

var _ansiEscapes2 = _interopRequireDefault(_ansiEscapes);

var _stringLength = require('string-length');

var _stringLength2 = _interopRequireDefault(_stringLength);

var _stripAnsi = require('strip-ansi');

var _stripAnsi2 = _interopRequireDefault(_stripAnsi);

var _Prompt = require('../shared/Prompt');

var _Prompt2 = _interopRequireDefault(_Prompt);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var pluralize = function pluralize(count, text) {
  return count === 1 ? text : `${text}s`;
};

var printPatternMatches = exports.printPatternMatches = function printPatternMatches(count, entity, pipe) {
  var extraText = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '';

  var pluralized = pluralize(count, entity);
  var result = count ? `\n\n Pattern matches ${count} ${pluralized}` : `\n\n Pattern matches no ${pluralized}`;

  pipe.write(result + extraText);
};

var printPatternCaret = exports.printPatternCaret = function printPatternCaret(pattern, pipe) {
  var inputText = `${_chalk2.default.dim(' pattern \u203A')} ${pattern}`;

  pipe.write(_ansiEscapes2.default.eraseDown);
  pipe.write(inputText);
  pipe.write(_ansiEscapes2.default.cursorSavePosition);
};

var printRestoredPatternCaret = exports.printRestoredPatternCaret = function printRestoredPatternCaret(pattern, currentUsageRows, pipe) {
  var inputText = `${_chalk2.default.dim(' pattern \u203A')} ${pattern}`;

  pipe.write(_ansiEscapes2.default.cursorTo((0, _stringLength2.default)(inputText), currentUsageRows - 1));
  pipe.write(_ansiEscapes2.default.cursorRestorePosition);
};

var printStartTyping = exports.printStartTyping = function printStartTyping(entity, pipe) {
  pipe.write(`\n\n ${_chalk2.default.italic.yellow(`Start typing to filter by a ${entity} regex pattern.`)}`);
};

var printMore = exports.printMore = function printMore(entity, pipe, more) {
  pipe.write(`\n   ${_chalk2.default.dim(`...and ${more} more ${pluralize(more, entity)}`)}`);
};

var printTypeaheadItem = exports.printTypeaheadItem = function printTypeaheadItem(item, pipe) {
  return pipe.write(`\n ${_chalk2.default.dim('\u203A')} ${item}`);
};

var formatTypeaheadSelection = exports.formatTypeaheadSelection = function formatTypeaheadSelection(item, index, activeIndex, prompt) {
  if (index === activeIndex) {
    prompt.setTypheadheadSelection((0, _stripAnsi2.default)(item));
    return _chalk2.default.black.bgYellow((0, _stripAnsi2.default)(item));
  }
  return item;
};
